src/coreclr/jit/lsraarm64.cpp

```diff
@@ -25,6 +25,247 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+
+    // First refposition should always start with RefTypeUse
+    assert(firstRefPosition->refType != RefTypeUpperVectorRestore);
+
+    INDEBUG(int refPosCount = 1);
+    regMaskTP busyConsecutiveRegMask = ~(((1ULL << firstRefPosition->regCount) - 1) << firstRegAssigned);
+
+    while (consecutiveRefPosition != nullptr)
+    {
+        assert(consecutiveRefPosition->regCount == 0);
+#if FEATURE_PARTIAL_SIMD_CALLEE_SAVE
+        if (consecutiveRefPosition->refType == RefTypeUpperVectorRestore)
+        {
+            if (consecutiveRefPosition->getInterval()->isPartiallySpilled)
+            {
+                // Make sure that restore doesn't get one of the registers that are part of series we are trying to set
+                // currently.
+                // TODO-CQ: We could technically assign RefTypeUpperVectorRestore and its RefTypeUse same register, but
+                // during register selection, it might get tricky to know which of the busy registers are assigned to
+                // RefTypeUpperVectorRestore positions of corresponding variables for which (another criteria)
+                // we are trying to find consecutive registers.
+
+                consecutiveRefPosition->registerAssignment &= ~busyConsecutiveRegMask;
+            }
+            consecutiveRefPosition = getNextConsecutiveRefPosition(consecutiveRefPosition);
+        }
+#endif // FEATURE_PARTIAL_SIMD_CALLEE_SAVE
+        INDEBUG(refPosCount++);
+        assert(consecutiveRefPosition->refType == RefTypeUse);
+        consecutiveRefPosition->registerAssignment = genRegMask(regToAssign);
+        consecutiveRefPosition                     = getNextConsecutiveRefPosition(consecutiveRefPosition);
+        regToAssign                                = regToAssign == REG_FP_LAST ? REG_FP_FIRST : REG_NEXT(regToAssign);
+    }
+
+    assert(refPosCount == firstRefPosition->regCount);
+
+    return true;
+}
+
+//------------------------------------------------------------------------
+// areNextConsecutiveRegistersBusy: Starting with `regToAssign`, check if next
+//   `registersToCheck` are free or not.
+//
+// Arguments:
+//      - First refPosition of the series of consecutive registers.
+//    regToAssign     - Register assigned to the first refposition.
+//    registersCount  - Number of registers to check.
+//    registerType    - Type of register.
+//
+//  Returns:
+//      True if all the consecutive registers starting from `regToAssign` were free. Even if one
+//      of them is busy, returns false.
+//
+bool LinearScan::areNextConsecutiveRegistersFree(regNumber regToAssign, int registersCount, var_types registerType)
+{
+    for (int i = 0; i < registersCount; i++)
+    {
+        if (isRegInUse(regToAssign, registerType))
+        {
+            return false;
+        }
+        regToAssign = regToAssign == REG_FP_LAST ? REG_FP_FIRST : REG_NEXT(regToAssign);
+    }
```

Comment:
We could presumably simplify this down to not loop. I believe the following are correct given the general algorithm for finding consecutively set bits:

To find the two consecutive bits:
```cpp
x &= (x >> 1);
```

To find three consecutive bits:
```cpp
x &= (x >> 1);
x &= (x >> 1);
```

To find four consecutive bits:
```cpp
x &= (x >> 2);
x &= (x >> 1);
```

For each of these, `x != 0` means there exists a sequence of 2/3/4 bits and the remaining set bit is the start of the set sequence.

Consider finding 3 consecutive bits in `9952`, which is `0b0010_0110_1110_0000`. This will end up producing `32`, which is `0b0010_0000` and hence we know that there is a sequence of bits and it is `0b1110_0000`